src/lib/Generator/UniqueStringGenerator.php

```diff
@@ -0,0 +1,17 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\PersonalizationClient\Generator;
+
+final class UniqueStringGenerator implements UniqueStringGeneratorInterface
+{
+    public function generate(int $length): string
+    {
+        return substr(md5(uniqid()), 0, $length);
+    }
+}
```

Comment:
Your goal is to generate username and password which are cryptographically secure and not necessarily unique right? This implementation is not meet this requirement.

Implementation proposal:

```php
final class SecureUniqueStringGenerator implements UniqueStringGeneratorInterface
{
    private const DEFAULT_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!-.[]?*()";

    private string $alphabet;
    
    private int $alphabetLength;

    public function __construct(string $alphabet = self::DEFAULT_ALPHABET)
    {
        $this->alphabet = $alphabet;
        $this->alphabetLength = strlen($alphabet);
    }

    public function generate(int $length): string
    {
        $value = '';
        for ($i = 0; $i < $length; $i++) {
            $value .= $this->alphabet[random_int(0, $this->alphabetLength)];
        }

        return $value;
    }
}
```

ping @glye for confirmation

Reply:
Yes, uniqid should not be used for cryptographic purposes as the doc says: https://www.php.net/manual/en/function.uniqid 

What @adamwojs proposes is based on random_int which is secure in itself. This doesn't specify the minimum length. 10-12 is now considered the minimum safe length in general terms. If this is a one time thing, there's no reason to go that short, we could say 30 for instance.

A simpler method is what we recommend in the doc: `bin2hex(random_bytes(N))` which returns 2N of hex chars. That's not a full alphabet, so should be a little longer than passwords generated by Adam's generator.

Reply:
Doc: I mean the app security doc, see https://doc.ibexa.co/en/latest/guide/security_checklist/#app_secret

Reply:
In this scenario we would like to generate const. length string (max. accepted by rest API) so I prefer to keep relatively  wide characters set. :wink:

Reply:
I see now that this is used for a 10 char username and 30 char password:
https://github.com/ezsystems/ezrecommendation-client/pull/84/files#diff-a8605ac9cc256c5f6f6571ff12f7e8906ce2adc8092b841e548f366e664d8b75R19

That's fine with either Adam's method or bin2hex(random_bytes). With Adam's method I would perhaps skip the special symbols and use only alphanumerics, to avoid any confusion the symbols might cause.

Reply:
@adamwojs Agree that if 30 chars is a hard limit, then the wider charset is safer, of course. Yours is ca 3 times wider than hex, so it has much more entropy. 30 chars hex isn't bad either, though :)

Reply:
> Your goal is to generate username and password which are cryptographically secure...

This requirement would be true if we were to use the generated value for key generation or hashing. We don't.

IMO It's simply not a cryptographic context.

If you want to ensure sufficient entropy then you can simply pass the `random_bytes` result to `md5` (replacing the `uniqid` call).

Reply:
I agree that sending random_bytes through md5, given the 30 char limit, is as good as sending random_bytes through bin2hex. Still I would protest against this. md5 should simply not be used in a password context, because it introduces doubt. Static code scanners like SonarCloud could trigger on this. For the same reason, uniqid should not be used when generating passwords.

I would say generating passwords is a cryptographic context, but the argument against the current code doesn't depend on this. The current code might be likely to be fine in practical usage in this case. Still we should not do it like this, because it looks bad both for static code scanners, and for manual code audits. Every time this is encountered, the auditor has to carefully consider the case. Using better practice approaches is easy, so why not do it?